test/unit/test-custom-element.js

```diff
@@ -1853,333 +1826,162 @@ describes.realWin('CustomElement', {amp: true}, (env) => {
-      });
-
-      it('should NOT turn on when enters viewport but already laid out', () => {
-        stubInA4A(false);
-        const timerDelay = env.sandbox.spy(Services.timerFor(win), 'delay');
-        const toggle = env.sandbox.spy(element, 'toggleLoading');
-        element.viewportCallback(true);
-        element.layoutCount_ = 1;
-        clock.tick(1000);
-        expect(timerDelay).to.have.been.called;
-        expect(toggle).to.have.not.been.called;
-      });
-
-      it('should not start loading when measured if already in viewport', () => {
-        stubInA4A(false);
-        const toggle = env.sandbox.spy(element, 'toggleLoading');
-        element.isInViewport_ = true;
-        element.setAttribute('layout', 'fill');
-        container.appendChild(element);
-        element.updateLayoutBox({top: 0, width: 300});
-        expect(toggle).to.not.be.called;
-      });
+      describe('toggleLoading with layout', () => {
+        let toggle;
 
-      it('should toggle loading off after layout complete', () => {
-        stubInA4A(false);
-        const toggle = env.sandbox.spy(element, 'toggleLoading');
-        element.setAttribute('height', '10');
-        element.setAttribute('width', '10');
-        container.appendChild(element);
-        return element.buildingPromise_
-          .then(() => {
-            return element.layoutCallback();
-          })
-          .then(() => {
-            expect(toggle).to.be.calledOnce;
-            expect(toggle.firstCall.args[0]).to.equal(false);
-            expect(toggle.firstCall.args[1].cleanup).to.equal(true);
-          });
-      });
+        beforeEach(() => {
+          toggle = env.sandbox.spy(element, 'toggleLoading');
+        });
 
-      it('should toggle loading off after layout failed', () => {
-        stubInA4A(false);
-        const toggle = env.sandbox.spy(element, 'toggleLoading');
-        env.sandbox
-          .stub(element.implementation_, 'layoutCallback')
-          .callsFake(() => {
-            return Promise.reject();
-          });
-        element.setAttribute('height', '10');
-        element.setAttribute('width', '10');
-        container.appendChild(element);
-        return element.buildingPromise_
-          .then(() => {
-            return element.layoutCallback();
-          })
-          .then(
-            () => {
-              throw new Error('Must have failed.');
-            },
-            () => {
-              expect(toggle).to.be.calledOnce;
-              expect(toggle.firstCall.args[0]).to.equal(false);
-              expect(toggle.firstCall.args[1].cleanup).to.equal(true);
-            }
-          );
-      });
+        it('should toggle loading off after layout complete', () => {
+          element.setAttribute('height', '10');
+          element.setAttribute('width', '10');
+          container.appendChild(element);
+          return element.buildingPromise_
+            .then(() => {
+              toggle.resetHistory();
```

Comment:
why is `toggle.resetHistory` necessary?

Reply:
Some the tests above test specifically connection/build loading indicator. But here I wanted to exclusively test only the loading chain and how it works with the loading indicator. This is actually an old test. But the old test only checked that the loading indicator is eventually reset. But I wanted to test a bit more here: that it's first set and then reset. So instead of only testing the last value, I'm testing all toggles during the loading process.